contracts/zero-ex/contracts/src/features/TransformERC20.sol

```diff
@@ -0,0 +1,365 @@
+    {
+        return _transformERC20Private(
+            keccak256(msg.data),
+            msg.sender,
+            inputToken,
+            outputToken,
+            inputTokenAmount,
+            minOutputTokenAmount,
+            transformations
+        );
+    }
+
+    /// @dev Internal version of `transformERC20()`. Only callable from within.
+    /// @param callDataHash Hash of the ingress calldata.
+    /// @param taker The taker address.
+    /// @param inputToken The token being provided by the taker.
+    ///        If `0xeee...`, ETH is implied and should be provided with the call.`
+    /// @param outputToken The token to be acquired by the taker.
+    ///        `0xeee...` implies ETH.
+    /// @param inputTokenAmount The amount of `inputToken` to take from the taker.
+    /// @param minOutputTokenAmount The minimum amount of `outputToken` the taker
+    ///        must receive for the entire transformation to succeed.
+    /// @return outputTokenAmount The amount of `outputToken` received by the taker.
+    function _transformERC20(
+        bytes32 callDataHash,
+        address payable taker,
+        IERC20TokenV06 inputToken,
+        IERC20TokenV06 outputToken,
+        uint256 inputTokenAmount,
+        uint256 minOutputTokenAmount,
+        Transformation[] memory transformations
+    )
+        public
+        override
+        payable
+        onlySelf
+        returns (uint256 outputTokenAmount)
+    {
+        return _transformERC20Private(
+            callDataHash,
+            taker,
+            inputToken,
+            outputToken,
+            inputTokenAmount,
+            minOutputTokenAmount,
+            transformations
+        );
+    }
+
+    /// @dev Private version of `transformERC20()`.
+    /// @param callDataHash Hash of the ingress calldata.
+    /// @param taker The taker address.
+    /// @param inputToken The token being provided by the taker.
+    ///        If `0xeee...`, ETH is implied and should be provided with the call.`
+    /// @param outputToken The token to be acquired by the taker.
+    ///        `0xeee...` implies ETH.
+    /// @param inputTokenAmount The amount of `inputToken` to take from the taker.
+    /// @param minOutputTokenAmount The minimum amount of `outputToken` the taker
+    ///        must receive for the entire transformation to succeed.
+    /// @return outputTokenAmount The amount of `outputToken` received by the taker.
+    function _transformERC20Private(
+        bytes32 callDataHash,
+        address payable taker,
+        IERC20TokenV06 inputToken,
+        IERC20TokenV06 outputToken,
+        uint256 inputTokenAmount,
+        uint256 minOutputTokenAmount,
+        Transformation[] memory transformations
+    )
+        public
+        payable
+        returns (uint256 outputTokenAmount)
+    {
+        // If the input token amount is -1, transform the taker's entire
+        // spendable balance.
+        if (inputTokenAmount == uint256(-1)) {
```

Comment:
We should also document this special value in the description, both in this fn and above.